File: lib/teleterm/services/userpreferences/userpreferences_test.go
// Teleport
// Copyright (C) 2023  Gravitational, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package userpreferences

import (
	"context"
	"testing"

	"github.com/stretchr/testify/require"

	userpreferencesv1 "github.com/gravitational/teleport/api/gen/proto/go/userpreferences/v1"
	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
)

var rootPreferencesMock = &userpreferencesv1.UserPreferences{
	Onboard: nil,
	Theme:   userpreferencesv1.Theme_THEME_LIGHT,
	ClusterPreferences: &userpreferencesv1.ClusterUserPreferences{
		PinnedResources: &userpreferencesv1.PinnedResourcesUserPreferences{
			ResourceIds: []string{"abc", "def"},
		},
	},
	UnifiedResourcePreferences: &userpreferencesv1.UnifiedResourcePreferences{
		DefaultTab:            userpreferencesv1.DefaultTab_DEFAULT_TAB_ALL,
		ViewMode:              userpreferencesv1.ViewMode_VIEW_MODE_CARD,
		LabelsViewMode:        userpreferencesv1.LabelsViewMode_LABELS_VIEW_MODE_COLLAPSED,
		AvailableResourceMode: userpreferencesv1.AvailableResourceMode_AVAILABLE_RESOURCE_MODE_NONE,
	},
}

var leafPreferencesMock = &userpreferencesv1.UserPreferences{
	Onboard: nil,
	ClusterPreferences: &userpreferencesv1.ClusterUserPreferences{
		PinnedResources: &userpreferencesv1.PinnedResourcesUserPreferences{
			ResourceIds: []string{"ghi", "jkl"},
		},
	},
}

func TestUserPreferencesGet(t *testing.T) {
	mockedRootClient := &mockClient{preferences: rootPreferencesMock}
	mockedLeafClient := &mockClient{preferences: leafPreferencesMock}

	response, err := Get(t.Context(), mockedRootClient, mockedLeafClient)
	require.NoError(t, err)
	require.Equal(t, rootPreferencesMock.GetUnifiedResourcePreferences(), response.GetUnifiedResourcePreferences())
	require.Equal(t, leafPreferencesMock.GetClusterPreferences(), response.GetClusterPreferences())
}

func TestUserPreferencesUpdateForRoot(t *testing.T) {
	mockedClient := &mockClient{preferences: rootPreferencesMock}

	newPreferences := &api.UserPreferences{
		ClusterPreferences: &userpreferencesv1.ClusterUserPreferences{
			PinnedResources: &userpreferencesv1.PinnedResourcesUserPreferences{
				ResourceIds: []string{"foo", "bar"},
			},
		},
		UnifiedResourcePreferences: nil,
	}

	updatedPreferences, err := Update(t.Context(), mockedClient, nil, newPreferences)
	require.NoError(t, err)
	// ClusterPreferences field has been updated with the new value.
	require.Equal(t, newPreferences.ClusterPreferences, mockedClient.upsertCalledWith.ClusterPreferences)
	require.Equal(t, newPreferences.ClusterPreferences, updatedPreferences.ClusterPreferences)
	// UnifiedResourcePreferences field has not changed because it was nil in the new value.
	require.Equal(t, rootPreferencesMock.UnifiedResourcePreferences, mockedClient.upsertCalledWith.UnifiedResourcePreferences)
	require.Equal(t, rootPreferencesMock.UnifiedResourcePreferences, updatedPreferences.UnifiedResourcePreferences)
	// Other user preferences have not been touched.
	require.Equal(t, rootPreferencesMock.Theme, mockedClient.upsertCalledWith.Theme)
}

func TestUserPreferencesUpdateForRootWithNoExistingPinnedResources(t *testing.T) {
	rootPreferences := &userpreferencesv1.UserPreferences{
		Theme: userpreferencesv1.Theme_THEME_LIGHT,
		ClusterPreferences: &userpreferencesv1.ClusterUserPreferences{
			PinnedResources: nil,
		},
		UnifiedResourcePreferences: rootPreferencesMock.UnifiedResourcePreferences,
	}
	mockedClient := &mockClient{preferences: rootPreferences}

	newPreferences := &api.UserPreferences{
		ClusterPreferences: &userpreferencesv1.ClusterUserPreferences{
			PinnedResources: &userpreferencesv1.PinnedResourcesUserPreferences{
				ResourceIds: []string{"foo", "bar"},
			},
		},
	}

	updatedPreferences, err := Update(t.Context(), mockedClient, nil, newPreferences)
	require.NoError(t, err)

	require.Equal(t, newPreferences.ClusterPreferences, mockedClient.upsertCalledWith.ClusterPreferences)
	require.Equal(t, newPreferences.ClusterPreferences, updatedPreferences.ClusterPreferences)
	require.Equal(t, rootPreferencesMock.UnifiedResourcePreferences, mockedClient.upsertCalledWith.UnifiedResourcePreferences)
	require.Equal(t, userpreferencesv1.Theme_THEME_LIGHT, mockedClient.upsertCalledWith.Theme)
}

func TestUserPreferencesUpdateForRootAndLeaf(t *testing.T) {
	mockedRootClient := &mockClient{preferences: rootPreferencesMock}
	mockedLeafClient := &mockClient{preferences: leafPreferencesMock}

	newPreferences := &api.UserPreferences{
		ClusterPreferences: &userpreferencesv1.ClusterUserPreferences{
			PinnedResources: &userpreferencesv1.PinnedResourcesUserPreferences{
				ResourceIds: []string{"foo", "bar"},
			},
		},
		UnifiedResourcePreferences: &userpreferencesv1.UnifiedResourcePreferences{
			DefaultTab:            userpreferencesv1.DefaultTab_DEFAULT_TAB_PINNED,
			ViewMode:              userpreferencesv1.ViewMode_VIEW_MODE_LIST,
			LabelsViewMode:        userpreferencesv1.LabelsViewMode_LABELS_VIEW_MODE_EXPANDED,
			AvailableResourceMode: userpreferencesv1.AvailableResourceMode_AVAILABLE_RESOURCE_MODE_REQUESTABLE,
		},
	}

	updatedPreferences, err := Update(t.Context(), mockedRootClient, mockedLeafClient, newPreferences)
	require.NoError(t, err)
	// ClusterPreferences field has been updated with the leaf cluster value.
	require.Equal(t, updatedPreferences.ClusterPreferences, mockedLeafClient.upsertCalledWith.ClusterPreferences)
	require.Equal(t, newPreferences.ClusterPreferences, updatedPreferences.ClusterPreferences)
	// UnifiedResourcePreferences field has been updated with the root cluster value.
	require.Equal(t, updatedPreferences.UnifiedResourcePreferences, mockedRootClient.upsertCalledWith.UnifiedResourcePreferences)
	require.Equal(t, newPreferences.UnifiedResourcePreferences, updatedPreferences.UnifiedResourcePreferences)
	// Other user preferences have not been touched.
	require.Equal(t, rootPreferencesMock.Theme, mockedRootClient.upsertCalledWith.Theme)
}

func TestNilUserPreferencesUpdate(t *testing.T) {
	tests := []struct {
		name                       string
		leafPreferences            *userpreferencesv1.UserPreferences
		expectedClusterPreferences *userpreferencesv1.ClusterUserPreferences
	}{
		{
			name:                       "root",
			expectedClusterPreferences: rootPreferencesMock.ClusterPreferences,
		},
		{
			name:                       "leaf",
			leafPreferences:            leafPreferencesMock,
			expectedClusterPreferences: leafPreferencesMock.ClusterPreferences,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			ctx := t.Context()

			mockedRootClient := &mockClient{preferences: rootPreferencesMock}
			var leafClient Client
			var mockedLeafClient *mockClient
			if tt.leafPreferences != nil {
				mockedLeafClient = &mockClient{preferences: tt.leafPreferences}
				leafClient = mockedLeafClient
			}

			updatedPreferences, err := Update(ctx, mockedRootClient, leafClient, nil)
			require.NoError(t, err)

			require.Nil(t, mockedRootClient.upsertCalledWith)
			if mockedLeafClient != nil {
				require.Nil(t, mockedLeafClient.upsertCalledWith)
			}
			require.Equal(t, rootPreferencesMock.UnifiedResourcePreferences, updatedPreferences.UnifiedResourcePreferences)
			require.Equal(t, tt.expectedClusterPreferences, updatedPreferences.ClusterPreferences)
		})
	}
}

type mockClient struct {
	preferences      *userpreferencesv1.UserPreferences
	upsertCalledWith *userpreferencesv1.UserPreferences
}

func (m *mockClient) GetUserPreferences(ctx context.Context, req *userpreferencesv1.GetUserPreferencesRequest) (*userpreferencesv1.GetUserPreferencesResponse, error) {
	return &userpreferencesv1.GetUserPreferencesResponse{
		Preferences: m.preferences,
	}, nil
}

func (m *mockClient) UpsertUserPreferences(ctx context.Context, req *userpreferencesv1.UpsertUserPreferencesRequest) error {
	m.upsertCalledWith = req.Preferences
	return nil
}

File: web/packages/teleport/src/services/userPreferences/userPreferences.test.ts
/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { SideNavDrawerMode } from 'gen-proto-ts/teleport/userpreferences/v1/sidenav_preferences_pb';
import { Theme } from 'gen-proto-ts/teleport/userpreferences/v1/theme_pb';
import { UserPreferences } from 'gen-proto-ts/teleport/userpreferences/v1/userpreferences_pb';

import {
  BackendUserPreferences,
  convertBackendUserPreferences,
  convertUserPreferences,
  isBackendUserPreferences,
} from 'teleport/services/userPreferences/userPreferences';
import {
  getCurrentTheme,
  getNextTheme,
  updateFavicon,
} from 'teleport/ThemeProvider';

test('should convert the old cluster user preferences format to the new one', () => {
  // this is how the backend currently returns cluster preferences - as an array of strings
  // instead of the protobuf representation of an object with a `resourceIds` field that contains
  // that array of strings
  const oldBackendPreferences: BackendUserPreferences = {
    keyboardLayout: 0,
    theme: Theme.LIGHT,
    clusterPreferences: {
      pinnedResources: ['resource1', 'resource2'],
    },
    sideNavDrawerMode: SideNavDrawerMode.COLLAPSED,
  };

  const actualUserPreferences: UserPreferences = {
    keyboardLayout: 0,
    theme: Theme.LIGHT,
    clusterPreferences: {
      pinnedResources: { resourceIds: ['resource1', 'resource2'] },
    },
    sideNavDrawerMode: SideNavDrawerMode.COLLAPSED,
  };

  // when we grab the user preferences from the local storage, we check if it is in the old format
  expect(isBackendUserPreferences(oldBackendPreferences)).toBe(true);
  expect(isBackendUserPreferences(actualUserPreferences)).toBe(false);

  // and convert it to the new format if it is
  const newPreferences = convertBackendUserPreferences(oldBackendPreferences);

  expect(newPreferences.clusterPreferences.pinnedResources.resourceIds).toEqual(
    oldBackendPreferences.clusterPreferences.pinnedResources
  );
});

test('should convert the user preferences back to the old format when updating', () => {
  // the backend still expects the old format when updating user preferences

  const actualUserPreferences: UserPreferences = {
    theme: Theme.LIGHT,
    keyboardLayout: 0,
    clusterPreferences: {
      pinnedResources: { resourceIds: ['resource1', 'resource2'] },
    },
    sideNavDrawerMode: SideNavDrawerMode.COLLAPSED,
  };

  const convertedPreferences = convertUserPreferences(actualUserPreferences);

  expect(convertedPreferences.clusterPreferences.pinnedResources).toEqual(
    actualUserPreferences.clusterPreferences.pinnedResources.resourceIds
  );
});

test('getCurrentTheme', () => {
  mockMatchMediaWindow('dark');
  let currentTheme = getCurrentTheme(Theme.UNSPECIFIED);
  expect(currentTheme).toBe(Theme.DARK);

  mockMatchMediaWindow('light');
  currentTheme = getCurrentTheme(Theme.UNSPECIFIED);
  expect(currentTheme).toBe(Theme.LIGHT);

  currentTheme = getCurrentTheme(Theme.LIGHT);
  expect(currentTheme).toBe(Theme.LIGHT);

  currentTheme = getCurrentTheme(Theme.DARK);
  expect(currentTheme).toBe(Theme.DARK);
});

describe('updateFavicon', () => {
  let originalMatchMedia: typeof window.matchMedia;

  beforeAll(() => {
    originalMatchMedia = window.matchMedia;
  });

  afterAll(() => {
    window.matchMedia = originalMatchMedia;
  });

  beforeEach(() => {
    document.body.innerHTML = '';
    const link = document.createElement('link');
    link.rel = 'icon';
    link.href = '/initial-favicon.png';
    document.head.appendChild(link);
  });

  test('set dark favicon when dark theme is preferred', () => {
    window.matchMedia = jest.fn().mockImplementation(query => ({
      matches: query === '(prefers-color-scheme: dark)',
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    }));

    updateFavicon();

    const favicon = document.querySelector(
      'link[rel="icon"]'
    ) as HTMLLinkElement;
    expect(favicon.href).toContain('/app/favicon-dark.png');
  });

  test('set light favicon when light theme is preferred', () => {
    window.matchMedia = jest.fn().mockImplementation(query => ({
      matches: query !== '(prefers-color-scheme: dark)',
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    }));

    updateFavicon();

    const favicon = document.querySelector(
      'link[rel="icon"]'
    ) as HTMLLinkElement;
    expect(favicon.href).toContain('/app/favicon-light.png');
  });
});

test('getNextTheme', () => {
  mockMatchMediaWindow('dark');
  let nextTheme = getNextTheme(Theme.UNSPECIFIED);
  expect(nextTheme).toBe(Theme.LIGHT);

  mockMatchMediaWindow('light');
  nextTheme = getNextTheme(Theme.UNSPECIFIED);
  expect(nextTheme).toBe(Theme.DARK);

  nextTheme = getNextTheme(Theme.LIGHT);
  expect(nextTheme).toBe(Theme.DARK);

  nextTheme = getNextTheme(Theme.DARK);
  expect(nextTheme).toBe(Theme.LIGHT);
});

function mockMatchMediaWindow(prefers: 'light' | 'dark') {
  return Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: query === `(prefers-color-scheme: ${prefers})`,
      media: query,
    })),
  });
}
